design/open/pull-request.md

```diff
@@ -0,0 +1,28 @@
+# Pull-request Proposal
+
+Enable a way to submit changes to a branch by creating a request to merge changes.
+The pull-request captures the merge operation we request to apply.
+
+### Goals
+
+- Capture request to merge a branch or specific reference
+- Mechanism to review, discuss and approve changes before applying to a branch
+- Automation using actions can verify / provide feedback on pull-request
+
+### How it will work
+
+Introduce a new resource called pull-request at the repository level.
+The pull-request will hold owner (the user who created the PR), assignee, reviewers the source branch/reference and the target branch.
+Each reviewer can set approved/reject (or clear his response) on the PR.
+PR can be merge operation is enabled when at least one reviewer approves or when no reviewers are assigned.
+PR can be closed at any time.
+PR will not be updated when new commits into the source branch.
+Merge PR will perform a merge operation and merge the PR as merged. If a conflict or an error occurs while merging, it will be returned as a response to the request.
+When a repository is deleted, all pull-requests will be deleted with it.
+
```

Comment:
How would branch protection rules be updated to support pull requests?

Reply:
Current branch protection enable merge only changes - as PR performs just merge we can use it to push changes into a branch just by merge/PRs.

Additional support can be added to the branch protection by allow only through PRs. And more options like - only approved, with specific checks and etc.